### PR TITLE
added basic e2e test for pipeline run command

### LINKED ISSUE #1369: Basic end to end test for pipeline run command
We merged a basic implementation of `kpt pipeline run` in last sprint. We need a basic test that tests end to end workflow for `pipeline run` command.

At the very basic, it starts with a simple kpt pkg, invokes the pipeline run command and compares the output to expected output.

### PR DESCRIPTION
**Pull request by natasha41575**
Close #1369 

Adds the most basic e2e test for pipeline run

Note: Currently pipeline run leaves behind a `builtin/setter-1: 'test'` annotation in all the yaml files, I'm not sure if this is expected or desired

Edited to include what I wrote in a comment later in the thread:
Based on our discussion today, I've updated the PR to include the test harness from kpt-functions-catalog that Donny wrote with a small tweak to allow it to work for pipeline. I've also included a basic pipeline run command test here. The kpt-functions-catalog repo can import and reuse this test harness when it is ready.

### CODE REVIEW

File: e2e/e2e_test.go
Comment (droot) [inline e2e/e2e_test.go:463]: I am still thinking how we would like to structure these tests. We will have to do some work upfront to create a test harness that does the following:
 1. Takes input pkg (either some structured format specified in the `Go` test files or from a directory under `testData` on filesystem)
 2. A way to specify corresponding expected output (either some form in `Go` file or under a directory say `golden` under `testData`)
 3. Runs `pipeline run` how we are doing today where setting up the input dirs etc and then comparing the output to the golden files. I remember seeing utility method about comparing the two pkgs to be Equal or not in one of @mortent PR. 
 
Think, if we were to test pkgs with hierarchy that has sub-packages etc., how we can setup multiple such tests and invoke it with different pipeline setup.

**Comment by droot** [comment]
> Note: Currently pipeline run leaves behind a builtin/setter-1: 'test' annotation in all the yaml files, I'm not sure if this is expected or desired

Yes, currently it's hardcoded, but Donny is working on a task that will enable running functions specified in the `pipeline.yaml` so that's another reason flexible test setup is important because expected output will depend on the content of the `pipeline.yaml`.

File: e2e/e2e_test.go
Comment (natasha41575) [reply e2e/e2e_test.go:463]: That makes sense, thanks. I will give more thought to this and would love to sync up if/when you have ideas for how these tests should be structured.

Comment (natasha41575) [reply e2e/e2e_test.go:471]: I've updated the test structure to use directories testData and golden under testutils

File: e2e/e2e_test.go
Comment (droot) [inline e2e/e2e_test.go:457]: Can the testData directory live in `pipeline` pkg itself ?

**Review by droot — COMMENTED** [review]
This is looking good. I have one minor question about the location of the testData.

File: e2e/e2e_test.go
Comment (frankfarzan) [reply e2e/e2e_test.go:471]: @mengqiy is setting up such e2e test harness with test packages that contain expected output/results.

File: e2e/e2e_test.go
Comment (mortent) [inline e2e/e2e_test.go:478]: Use subtests here.

File: e2e/e2e_test.go
Comment (mortent) [inline e2e/e2e_test.go:485]: Is using a golden directory structure for the output scalable here? Our experience with packaging is that there are many cases to test and that providing a large number of golden results is hard to maintain.

**Comment by mengqiy** [comment]
@natasha41575 In case you are curious about what we did in kpt-functions-catalog.
Our package is self-contained and it has everything needed for testing.
Let's use [kubeval](https://github.com/GoogleContainerTools/kpt-functions-catalog/tree/master/examples/kubeval) as an example:
```
.
├── .expected
│   ├── config.yaml
│   └── results.yaml
├── .krmignore
├── README.md
├── configs
│   └── example-config.yaml
└── fn-config.yaml
```
And https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/f4b9e3af8ee2ed405ba264691181f5fc90677367/tests/e2e-test/tests/e2e_test.go#L14-L31 describes what these files are.

**Comment by natasha41575** [comment]
@mengqiy Thank you so much! I will take a look and use that as a reference

**Comment by frankfarzan** [comment]
> @mengqiy Thank you so much! I will take a look and use that as a reference

I expect the functions catalog and kpt repo to share the same test harness for testing hydration pipeline. Is that the plan?

**Comment by Shell32-Natsu** [comment]
> > @mengqiy Thank you so much! I will take a look and use that as a reference
> 
> I expect the functions catalog and kpt repo to share the same test harness for testing hydration pipeline. Is that the plan?

Technically there is no blocker to use the same harness.

**Comment by natasha41575** [comment]
@frankfarzan We can make that the plan. I will talk to @droot to get his input

**Comment by mengqiy** [comment]
> > > @mengqiy Thank you so much! I will take a look and use that as a reference
> >
> > I expect the functions catalog and kpt repo to share the same test harness for testing hydration pipeline. Is that the plan?
> 
> Technically there is no blocker to use the same harness.

These 2 repos eventually should use the same test harness. I expect the primary use case of kpt function in the future will be through the hydration pipeline. Using the same test harness will make more sense. 
But it should not be a blocker that stop either project from moving forward at the moment.

@natasha41575 If you need my input or help with the test harness design, I'm happy to help.

**Comment by natasha41575** [comment]
@droot @Shell32-Natsu @mengqiy @frankfarzan 
Based on our discussion today, I've updated the PR to include the test harness from kpt-functions-catalog with a small tweak to allow it to work for pipeline. I've also included a basic pipeline run command test here. The kpt-functions-catalog repo can import and reuse this test harness when it is ready.

File: e2e/e2e_test.go
Comment (Shell32-Natsu) [inline e2e/e2e_test.go:40]: This comment needs to be updated

File: pkg/runner/config.go
Comment (Shell32-Natsu) [inline pkg/runner/config.go:1]: It's a little bit confusing to put these files under `pkg/runner`. Maybe `pkg/test/runner` or something.

File: pkg/runner/runner.go
Comment (Shell32-Natsu) [inline pkg/runner/runner.go:54]: The `results` dir is not used to store original package content. It's used to hold the kpt output when it has error. Is this changed in pipeline?

**Review by Shell32-Natsu — CHANGES_REQUESTED** [review]

File: pkg/runner/runner.go
Comment (mengqiy) [inline pkg/runner/runner.go:47]: `"kpt-fn-catalog-e2e-*"` this should be changed.

File: internal/pipeline/examples/basicpipeline/.expected/diff.patch
Comment (mengqiy) [inline internal/pipeline/examples/basicpipeline/.expected/diff.patch:8]: I don't see a pipeline definition that can cause these diff to happen.
Did I miss something?

Comment (Shell32-Natsu) [reply internal/pipeline/examples/basicpipeline/.expected/diff.patch:8]: There is a *builtin* function for testing. I think that's the reason.

Comment (natasha41575) [reply internal/pipeline/examples/basicpipeline/.expected/diff.patch:8]: Currently fnFilters is hardcoded to add these annotations. After #1388 they should disappear. Probably better to wait until that is merged before merging this one

File: pkg/runner/runner.go
Comment (natasha41575) [reply pkg/runner/runner.go:54]: I thought results dir was meant to originally meant for the functions to write their result to that dir. 
https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/127a1afe906e8b2a7d1d37fc14106b03d9631ca8/tests/e2etest/runner/runner.go#L70

In the original harness that you wrote, my understanding is that it runs kpt fn run on tmpPkgPath but stores the results in results dir and then compares them. 

kpt pipeline does not have a --results-dir option. The idea I had was to alter the harness so that it saves the original package in orgPkgPath, runs kpt <cmd> run on tmpPkgPath, and then compares them. 

If this makes sense I can change the comment and the name of the dir to make it more clear, but please correct me if I am misunderstanding something

File: e2e/e2e_test.go
Comment (natasha41575) [reply e2e/e2e_test.go]: updated

File: pkg/runner/config.go
Comment (natasha41575) [reply pkg/runner/config.go]: moved to pkg/test/runner

File: pkg/runner/runner.go
Comment (natasha41575) [reply pkg/runner/runner.go]: done

File: pkg/runner/runner.go
Comment (natasha41575) [reply pkg/runner/runner.go]: in pipeline there is no --results-dir argument so I changed it to transform tmpPkgPath and compare it against orgPkgPath. but it is different and that doesn't work for fn so I updated the PR to handle fn differently

**Review by droot — APPROVED** [review]
Awesome!
/lgtm

File: pkg/test/runner/config.go
Comment (mengqiy) [inline pkg/test/runner/config.go:27]: We default RunTimes to 2 for functions catalog to ensure idempotency.
Does 2 still make sense for pipeline use case?

File: pkg/test/runner/config.go
Comment (mengqiy) [inline pkg/test/runner/config.go:40]: Same question here

**Review by mengqiy — COMMENTED** [review]
1 minor comment, otherwise LGTM

**Comment by frankfarzan** [comment]
> @droot @Shell32-Natsu @mengqiy @frankfarzan
> Based on our discussion today, I've updated the PR to include the test harness from kpt-functions-catalog that Donny wrote with a small tweak to allow it to work for pipeline. I've also included a basic pipeline run command test here. The kpt-functions-catalog repo can import and reuse this test harness when it is ready.

File: pkg/test/runner/config.go
Comment (natasha41575) [reply pkg/test/runner/config.go]: Talked to Sunil and this is not a requirement for pipeline. As a result I've changed the default to 1, thanks for pointing this out. But that means on your functions-catalog tests you will have to specify your tests to run twice when you need to check idempotency

File: e2e/e2e_test.go
Comment (Shell32-Natsu) [inline e2e/e2e_test.go:55]: The runner type probably should be passed when we create the runner (`runner.NewRunner(c)`). And it will be good to have a const value and type defined instead of strings.

File: pkg/test/runner/runner.go
Comment (Shell32-Natsu) [inline pkg/test/runner/runner.go:145]: This will make the test fail directly if there is an error. What if an error is expected?

File: e2e/e2e_test.go
Comment (natasha41575) [reply e2e/e2e_test.go]: NewRunner now takes a second argument for the command. And I defined consts for the command names

File: pkg/test/runner/runner.go
Comment (natasha41575) [reply pkg/test/runner/runner.go]: changed it so that the error gets returned and test for pipeline can be written to expect errors

**Comment by natasha41575** [comment]
@Shell32-Natsu please take another look when you get the chance, thank you

**Review by Shell32-Natsu — APPROVED** [review]
lgtm

**Pull request merged**
